src/core/portal/portal.ts

```diff
@@ -0,0 +1,194 @@
+import {ElementRef} from 'angular2/core';
+import {ComponentRef} from 'angular2/core';
+
+import {BaseException} from 'angular2/src/facade/exceptions';
+import {PromiseWrapper} from 'angular2/src/facade/promise';
+
+
+/**
+ * A `Portal` is something that you want to render somewhere else.
+ * It can be attach to / detached from a `PortalHost`.
+ */
+export abstract class Portal<T> {
+  private attachedHost_: PortalHost;
+
+  /** Attach this portal to a host. */
+  attach(host: PortalHost): Promise<T> {
+    if (host == null) {
+      throw new BaseException('Attempting to attach a portal to a null host');
+    }
+
+    if (host.hasAttached()) {
+      throw new BaseException('Host already has a portal attached');
+    }
+
+    this.attachedHost_ = host;
+    return <Promise<T>> host.attach(this);
+  }
+
+  /** Detach this portal from its host */
+  detach(): Promise<void> {
+    let host = this.attachedHost_;
+    if (host == null) {
+      throw new BaseException('Portal has no host from which to detach');
+    }
+
+    this.attachedHost_ = null;
+    return host.detach();
+  }
+
+  /** Whether this portal is attached to a host. */
+  get isAttached(): boolean {
+    return this.attachedHost_ != null;
+  }
+
+  /**
+   * Sets the PortalHost reference without performing `attach()`. This is used directly by
+   * the PortalHost when it is performing an `attach()` or `detatch()`.
+   */
+  setAttachedHost(host: PortalHost) {
+    this.attachedHost_ = host;
+  }
+}
+
+
+/**
+ * A `ComponentPortal` is a portal that instantiates some Component upon attachment.
+ */
+export class ComponentPortal extends Portal<ComponentRef> {
+  /** The type of the component that will be instantiated for attachment. */
+  public component: Type;
+
+  /**
+   * [Optional] Where the attached component should live in Angular's *logical* component tree.
+   * This is different from where the component *renders*, which is determined by the PortalHost.
+   * The origin necessary when the host is outside of the Angular application context.
+   */
+  public origin: ElementRef;
+
+  constructor(component: Type, origin: ElementRef = null) {
+    super();
+    this.component = component;
+    this.origin = origin;
+  }
+}
+
+
+/**
+ * A `TemplatePortal` is a portal that represents some embedded template (TemplateRef).
+ */
+export class TemplatePortal extends Portal<Map<string, any>> {
+  /** The embedded template that will be used to instantiate an embedded View in the host. */
+  templateRef: TemplateRef;
+
+  /** Additional locals for the instantiated embedded view. */
+  locals: Map<string, any> = new Map<string, any>();
+
+  constructor(template: TemplateRef) {
+    super();
+    this.templateRef = template;
+  }
+
+  get origin(): ElementRef {
+    return this.templateRef.elementRef;
+  }
+
+  attach(host: PortalHost, locals?: Map<string, any>): Promise<Map<string, any>> {
+    this.locals = locals == null ? new Map<string, any>() : locals;
```

Comment:
Why change it? Using `== null` is idomatic TS / JS that has the wonderfully rare benefit of also working in Dart.